contract/models/contract.py

```diff
@@ -222,72 +221,103 @@ def _prepare_invoice_line(self, line, invoice_id):
+            contract.company_id.currency_id
         )
         invoice = self.env['account.invoice'].new({
-            'reference': self.code,
+            'reference': ', '.join(self.filtered('code').mapped('code')),
             'type': 'out_invoice',
-            'partner_id': self.partner_id.address_get(
+            'partner_id': contract.partner_id.address_get(
                 ['invoice'])['invoice'],
             'currency_id': currency.id,
             'journal_id': journal.id,
-            'date_invoice': self.recurring_next_date,
-            'origin': self.name,
-            'company_id': self.company_id.id,
-            'contract_id': self.id,
-            'user_id': self.partner_id.user_id.id,
+            'date_invoice': contract.recurring_next_date,
+            'origin': ', '.join(self.mapped('name')),
+            'company_id': contract.company_id.id,
+            'contract_id': contract.id,
+            'user_id': contract.partner_id.user_id.id,
         })
         # Get other invoice values from partner onchange
         invoice._onchange_partner_id()
         return invoice._convert_to_write(invoice._cache)
 
     @api.multi
     def _create_invoice(self):
-        self.ensure_one()
+        """Create the invoice from the source contracts.
+
+        :param self: Contract records. Invoice header data will be obtained
+          from the first record of this recordset.
+
+        :return Created invoice record.
+        """
         invoice_vals = self._prepare_invoice()
         invoice = self.env['account.invoice'].create(invoice_vals)
-        for line in self.recurring_invoice_line_ids:
-            invoice_line_vals = self._prepare_invoice_line(line, invoice.id)
-            self.env['account.invoice.line'].create(invoice_line_vals)
-        invoice.compute_taxes()
-        return invoice
-
-    @api.multi
-    def recurring_create_invoice(self):
-        invoices = self.env['account.invoice']
         for contract in self:
             old_date = fields.Date.from_string(
-                contract.recurring_next_date or fields.Date.today())
+                contract.recurring_next_date or fields.Date.today(),
+            )
             new_date = old_date + self.get_relalive_delta(
-                contract.recurring_rule_type, contract.recurring_interval)
+                contract.recurring_rule_type, contract.recurring_interval,
+            )
             ctx = self.env.context.copy()
             ctx.update({
                 'old_date': old_date,
                 'next_date': new_date,
-                # Force company for correct evaluate domain access rules
+                # For correct evaluating of domain access rules + properties
                 'force_company': contract.company_id.id,
             })
-            # Re-read contract with correct company
-            invoices |= contract.with_context(ctx)._create_invoice()
+            obj = self.with_context(ctx)
```

Comment:
In general I think it's easier to use kwargs, which include the previous context and just override the supplied keys.

    obj = self.with_context(old_date=old_date, new_date=new_date, force_company=contract.company_id.id)

The result is the same, so not a requirement 😸